doc/man3/X509_dup.pod

```diff
@@ -359,6 +359,16 @@ algorithms from providers. This created object can then be used when loading
 binary data using B<d2i_I<TYPE>>().
 
 B<I<TYPE>_dup>() copies an existing object, leaving it untouched.
+Note, however, that the internal representation of the object
+may contain extra data besides the ASN.1 structure.
+For instance, an B<X509> object usually is augmented
+by cached information on X.509v3 extensions, etc.
+Such extra data is not copied.
+If it needs to be retained, it may be better to use B<I<TYPE>_up_ref>().
+For the case of B<X509> objects, an alternative to using L<X509_up_ref(3)>
+is to still call B<I<TYPE>_dup>(), e.g., I<copied_cert = X509_dup(cert)>,
+followed by I<X509_check_purpose(copied_cert, -1, 0)>,
+which re-builds the cached data.
```

Comment:
This does not explain whether the cached data is anything more than an optimisation. Is there any *harm* if it is not copied?

One might also note that computing the cached data is a mutation, and should not be done for objects that are visible in more than one thread, so in cases where the X509 object might later be mutated (including via caching of various computed properties), a *dup* is preferable to an upref.

We should I think unconditionally recommend upref over dup.

Reply:
Very good points!

Indeed there is harm if the cached data is not copied. I came across this issue analyzing and fixing an obscure cert verification bug in libSecUtils when CRLs are involved: https://github.com/siemens/libsecutils/pull/57
In that case, the missing cached data causes the check in `x509_vfy.c` following line 1311:
```
/* Check cert for matching CRL distribution points */
```
to go wrong because the cert does contain CDP data, yet this information is missing from the (falsely empty) cached field `x->crldp`. This in turn leads to the misleading error `X509_V_ERR_DIFFERENT_CRL_SCOPE`.

I've now made clear that things can go wrong when just using `TYPE_dup()`
and now more strongly recommend using `TYPE_up_ref()`.
We cannot generally recommend it, also because it is not available for all types.

I've also rephrased 'extra data' to 'further data'  in order to avoid the potential confusion you mentioned.

Reply:
I see this was already merged, but I am curious why it is that absence of lazily-computed data causes observable behaviour differences. If that's the case, perhaps the real issue is not in documentation, but API semantics.

When the computed data is missing, the relevant thread should lock the object and recompute the missing data.  Or else, the data should automatically be computed as part of X509_dup().

Why should users need to know about these sort of rather subtle behaviour differences?

Reply:
Perhaps we should ensure that X509_dup() and/or CRL_dup() always compute the auxiliary cached information, so it is always pre-computed as part of decoding a certificate, this is perhaps something that should happen at the d2i layer. And then dup can just be i2d, followed by d2i or equivalent.

I am not fond of having to document warts...

Reply:
Sounds good - please continue the discussion on #23337 which I opened for this purpose as this PR is closed.